crypto/rsa_extra/rsassa_pss_asn1.c

```diff
@@ -0,0 +1,317 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <openssl/evp.h>
+
+#include <openssl/bytestring.h>
+#include <openssl/digest.h>
+#include <openssl/mem.h>
+#include <openssl/err.h>
+
+#include "../internal.h"
+#include "rsassa_pss.h"
+
+const RSA_PSS_SUPPORTED_ALGOR sha1_func = {
+    NID_sha1,
+    // 1.3.14.3.2.26
+    {0x2b, 0x0e, 0x03, 0x02, 0x1a},
+    5,
+};
+
+const RSA_PSS_SUPPORTED_ALGOR sha224_func = {
+    NID_sha224,
+    // 2.16.840.1.101.3.4.2.4
+    {0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x04},
+    9,
+};
+
+const RSA_PSS_SUPPORTED_ALGOR sha256_func = {
+    NID_sha256,
+    // 2.16.840.1.101.3.4.2.1
+    {0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01},
+    9,
+};
+
+const RSA_PSS_SUPPORTED_ALGOR sha384_func = {
+    NID_sha384,
+    // 2.16.840.1.101.3.4.2.2
+    {0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02},
+    9,
+};
+
+const RSA_PSS_SUPPORTED_ALGOR sha512_func = {
+    NID_sha512,
+    // 2.16.840.1.101.3.4.2.3
+    {0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03},
+    9,
+};
+
+// Used to check if the oid is in the supported five hash functions.
+// Section 2.1. https://tools.ietf.org/html/rfc4055#page-4
+static const RSA_PSS_SUPPORTED_ALGOR *const rsa_pss_hash_functions[] = {
+    &sha1_func, &sha224_func, &sha256_func, &sha384_func, &sha512_func,
+};
+
+const RSA_PSS_SUPPORTED_ALGOR MGF1 = {
+    NID_mgf1,
+    // 1.2.840.113549.1.1.8
+    {0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x08},
+    9,
+};
+
+// Used to check if the oid is in the supported mask generation functions.
+// Section 2.2. https://tools.ietf.org/html/rfc4055#page-4
+static const RSA_PSS_SUPPORTED_ALGOR *const rsa_pss_mg_functions[] = {
+    &MGF1,
+};
+
+// parse_oid return one on success and zero on failure.
+// Given |oid| and |supported_algors|, if match, set |*out| to be the algorithm
+// identifier.
+static int parse_oid(CBS *oid,
+                     const RSA_PSS_SUPPORTED_ALGOR *const supported_algors[],
+                     size_t size, RSA_ALGOR_IDENTIFIER **out) {
+  for (size_t i = 0; i < size; i++) {
+    const RSA_PSS_SUPPORTED_ALGOR *supported_algr = supported_algors[i];
+    if (CBS_len(oid) == supported_algr->oid_len &&
+        OPENSSL_memcmp(CBS_data(oid), supported_algr->oid,
+                       supported_algr->oid_len) == 0) {
+      *out = RSA_ALGOR_IDENTIFIER_new();
+      if (*out != NULL) {
+        (*out)->nid = supported_algr->nid;
+      }
+      return (*out) != NULL;
+    }
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_UNSUPPORTED_ALGORITHM);
+  return 0;
+}
+
+// For One-way Hash Functions:
+// All implementations MUST accept both NULL and absent parameters as legal and
+// equivalent encodings. See 2.1. https://tools.ietf.org/html/rfc4055#page-5
+static int is_absent_or_null(CBS *params) {
```

Comment:
nit: suggest changing name to indicate this is for parameters field.

Reply:
No change. The method comment and signature should be enough.